### ISSUE TITLE
MaskedLoss is confusing/incorrect

### ISSUE THREAD

**Issue by razorx89** [Project-MONAI/MONAI#4193]
The current implementation applies a user provided mask by multiplying the prediction and groundtruth with the supplied mask.

https://github.com/Project-MONAI/MONAI/blob/df177fd3d992f40d549c198c8a08addf7bd686c6/monai/losses/spatial_mask.py#L63

However, this is only correct in very specific cases like Dice with ignored background (first) class. Since `MaskedDiceLoss` passes all arguments unfiltered to the `DiceLoss` it is possible to break the mechanism by passing `ignore_background=False`. Furthermore, the `MaskedLoss` wrapper class can be used on any loss function and there it becomes tricky. In case of cross entropy, the total loss over all voxels is not correct. Masked or ignored voxels should not contribute to the loss function and therefore the number of voxels for computing the mean should be a subset of the total number of voxels. In its current form the CE/Focal losses, or the combinations with Dice, will compute a mean over all voxel losses, where maybe most of the voxels have zero losses.

For context: I stumbled across this implementation detail when trying to implement a masking for ignored voxels which are available in the segmentation file as label 255. Those were skipped during the annotation process and thus the segmentation files are sparsely annotated. Only on annotated, voxels which do not have a 255 label, I can compute the loss and the rest of the image is just available for providing the network context information.

One possible solution, or at least that is how I handle this case, would be to flatten the spatial dimensions as 1D (B, C, N) and index only non-masked elements. This requires additionally that each batch item is processed by the loss function individually, since each batch item can have a different number of masked elements. A nice positive benefit is that depending on the volume size the computational cost of the loss function is reduced (in my case only every fifth slice is annotated, so for 80% ignored voxels the loss function does not have to be computed). However, losses which explicitly take the spatial relationships into account would need a different handling (e.g. a Housdorff Loss) .

cc @nvahmadi

**Comment by wyli** [comment]
Agreed, as I can see that the `MaskedLoss` as a wrapper should have an `"indexing"` mode to mask out the spatial locations, in addition to the masking by element-wise multiplication. https://github.com/Project-MONAI/MONAI/blob/df177fd3d992f40d549c198c8a08addf7bd686c6/monai/losses/spatial_mask.py#L22

for `MaskedDiceLoss`, in my understanding, it provides an element-wise weighting, the weights could be non-binary (and maybe differentiable). could you provide a use case of incorrect results for this, I can think of how to improve it.

**Comment by razorx89** [comment]
I don't know if having a non-binary target would still be a valid Dice set formulation. If you have examples of this then please proof me wrong.

https://github.com/Project-MONAI/MONAI/blob/df177fd3d992f40d549c198c8a08addf7bd686c6/monai/losses/dice.py#L165-L174

For example: One voxel has a target of 0.25 and a prediction of 0.25, then the loss would not be 0. It would be `(0.25 * 0.25) / (0.25 + 0.25) = 0.125`.

I think the idea behind the multiplication may be something like class weighting, but this should be implemented differently. Each dice loss/score of the classes should be weighted according to provided weights. And if the intention was to individually weight cross entropy losses on the a per-voxel basis, then the weighting mask should be used during reduction.

**Comment by razorx89** [comment]
Thanks for the reference, I see your point. But in its current state equation 3 is not correctly implemented by MaskedDiceLoss, since in the intersection term we would end up with `w_ci^2`. I would further propose that there might be a class level separation between weighting and masking. That was the main point which confused me about the `MaskedLoss` class.

**Comment by jxchen01** [comment]
Any update on this thread? I don't quite understand how the MaskedLoss works for other losses like CrossEntropyLoss when the mask is not binary. 

For example: 

When computing`return self.loss(input * mask, target * mask)`, suppose our `mask` contains some entries with value == 100, otherwise == 1, and our `target` is from 0 to 9 (a 10-class segmentation problem). Then, `target * mask` will result in values of 1000. Then, if we pass such `target * mask` into the `self.loss()` calculation. I think this is wrong. It becomes a 1000-class segmentation problem, instead of 10.

**Comment by jxchen01** [comment]
I am not sure if my understanding is correct. But, i think no matter how, when pass `mask` with non-binary weight, e.g, weight = 1.25, then for a classification loss, `target * mask` does not contain "class" values anymore. e.g. `target = [0, 1, 1, 0] and mask = [0, 1.25, 1.25, 1]` then `target * mask` becomes `[0, 1.25, 1.25, 0]`, which is not a valid input for certain losses like CrossEntropy.

**Comment by dyollb** [comment]
It seems this issue was never really solved. Maybe [`torch.MaskedTensor`](https://pytorch.org/docs/stable/masked.html) would allow to implement this wrapper more generically.

**Comment by dancingreindeer** [comment]
> One possible solution, or at least that is how I handle this case, would be to flatten the spatial dimensions as 1D (B, C, N) and index only non-masked elements. This requires additionally that each batch item is processed by the loss function individually, since each batch item can have a different number of masked elements. A nice positive benefit is that depending on the volume size the computational cost of the loss function is reduced (in my case only every fifth slice is annotated, so for 80% ignored voxels the loss function does not have to be computed). However, losses which explicitly take the spatial relationships into account would need a different handling (e.g. a Housdorff Loss) .
> 
> cc @nvahmadi

Instead of flatten the spatial dimension as (B,C,N), could you flatten it as (1,C,BxN), where N is the total number of voxels in each batch item, and BxN represents the total voxels in your current batch. 

This will allow you to compute loss on all voxels through your indexing technique, while remaining the channel dimension. 

The MaskedLoss wrapper is not implemented as described by docstring in my opinion. The issue is that the "setting to 0" happened before softmax. I am not sure why this is not fixed yet?
